### ISSUE TITLE
Canceled invoice can be canceled again

### ISSUE THREAD

**Issue by redelschaap** [magento/magento2#9968]
A canceled invoice can be canceled again. I think `\Magento\Sales\Model\Order\Invoice::cancel` should check whether the invoice can be canceled before executing all cancel operations.

### Preconditions

1. Tested on Magento 2.1.5

### Steps to reproduce

1. Load a canceled invoice programmatically
2. Call `$invoice->cancel()`

### Expected result

1. Throw an exception saying the invoice has already been canceled.

### Actual result

1. The invoice gets canceled twice, so order amounts are also being adjusted twice.

**Comment by magento-engcom-team** [comment]
@redelschaap, thank you for your report.
We've created internal ticket(s) MAGETWO-75779 to track progress on the issue.

**Comment by osrecio** [comment]
Maybe Adding a Guard Clause in `\Magento\Sales\Model\Order\Invoice::cancel` with:

```
public function cancel()
{
        if (!$this->canCancel()) {
            return $this;
        }
       ...
}
```
can Solve the problem.

If you accept this solution @okorshenko @magento-engcom-team @veloraven  can proceed to create PR.

**Comment by okorshenko** [comment]
We need to backport this into 2.2-develop

**Comment by osrecio** [comment]
Working on it

**Comment by osrecio** [comment]
Created 2.2-develop Backport @okorshenko

**Comment by okorshenko** [comment]
Hi @redelschaap the issue has been fixed in 2.2-develop branch and will be available in 2.2.2 soon

**Comment by okorshenko** [comment]
Hi @redelschaap the issue has been fixed in 2.2-develop branch and will be available in 2.2.2 soon
